Day 3/solution2/src/main.rs

```diff
@@ -1,68 +1,23 @@
-use std::fs::File;
-use std::io::{self, BufRead};
-use std::path::Path;
-// use std::slice;
+use std::fs;
 
 fn main() {
-    // let mut counts : Vec<Vec<u32>> = vec![vec![0, 2]; 12];
-    let count: Vec<u32> = {
-        let lines: io::Lines<_> = read_lines("../input.txt").unwrap();
-        let lines_vector: Vec<String> = lines.map(|line| line.unwrap()).collect::<Vec<String>>();
-        let lines_count_vector: Vec<Vec<u32>> = lines_vector
-            .into_iter()
-            .map(|line| line.char_count('1'))
-            .collect();
-        let counts = lines_count_vector
-            .into_iter()
-            .reduce(|a, b| a.add(b))
-            .unwrap();
-        counts
-            .iter()
-            .map(|count| if *count >= 500 { 1 } else { 0 })
-            .collect()
-    };
+    let count: Vec<u32> = fs::read_to_string("../input.txt")
+        .unwrap()
+        .lines()
+        .map(|line| {
+            line.chars()
+                .map(|c| match c {
+                    '0' => 0,
+                    '1' => 1,
+                    _ => unreachable!(),
+                })
+                .collect::<Vec<u32>>()
+        })
+        .reduce(|a, b| a.iter().zip(b.iter()).map(|(&x, &y)| x + y).collect())
+        .unwrap()
+        .iter()
+        .map(|count| if *count >= 500 { 1 } else { 0 })
+        .collect();
 
     println!("{:?}", count)
 }
-
-fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
-where
-    P: AsRef<Path>,
-{
-    let file = File::open(filename)?;
-    Ok(io::BufReader::new(file).lines())
-}
-
-trait Addable {
-    fn add(&self, other: Self) -> Self;
-}
-
-impl Addable for Vec<u32> {
-    fn add(&self, other: Vec<u32>) -> Vec<u32> {
-        let mut new_vec: Vec<u32> = Vec::new();
-        for (v1, v2) in self.iter().zip(other) {
-            new_vec.push((*v1) + v2);
-        }
-        new_vec
-    }
-}
-
-trait Countable {
-    fn char_count(&self, character: char) -> Vec<u32>;
-}
-
-impl Countable for String {
-    fn char_count(&self, character: char) -> Vec<u32> {
-        let mut count: Vec<u32> = Vec::new();
```

Comment:
You definitely did a lot more work with these sections to get what you want out of them, but that's likely because it seemed like the right way to do it. Tbh this is more complex stuff that you tend to learn down the line, so it's great that you were exposed to it now :) adding functionality to types with traits and impls is super powerful, and is at the core of lots of soft eng in Rust. So congrats on getting it to work like this :partying_face: 

The version I wrote uses more built-in iterator functions that I'm familiar with, which is something that you gain intuition for over time. It's not super clear how to use them all together without the practice. It actually took me quite a while to get the reduce section working, I asked in the Veloren Discord for how to add all the vecs together. It's all about practice!